sigal/plugins/compress_assets.py

```
@@ -0,0 +1,157 @@
+        if not compressed_filename:
+            return
+
+        self.do_compress(filename, compressed_filename)
+
+    def can_compress(self, filename):
+        # Be sure it's a path
+        filename = Path(filename)
+        if not filename.suffix[1:] in self.settings['suffixes']:
+            return False
+
+        file_stats = None
+        compressed_stats = None
+        compressed_filename_result = self.compressed_filename(filename)
+        try:
+            file_stats = os.stat(filename)
+            compressed_stats = os.stat(compressed_filename_result)
+        except FileNotFoundError:
+            pass
+
+        if file_stats and compressed_stats:
+            return compressed_filename_result if file_stats.st_mtime > compressed_stats.st_mtime else False
+        else:
+            return compressed_filename_result
+
+
+class GZipCompressor(BaseCompressor):
+
+    def suffix(self):
+        return 'gz'
+
+    def do_compress(self, filename, compressed_filename):
+        with open(filename, 'rb') as f_in, gzip.open(compressed_filename, 'wb') as f_out:
+            shutil.copyfileobj(f_in, f_out)
+
+
+class ZopfliCompressor(BaseCompressor):
+
+    def suffix(self):
+        return 'gz'
+
+    def do_compress(self, filename, compressed_filename):
+        with open(filename, 'rb') as f_in, open(compressed_filename, 'wb') as f_out:
+            f_out.write(zopfli.gzip.compress(f_in.read()))
+
+
+class BrotliCompressor(BaseCompressor):
+
+    def suffix(self):
+        return 'br'
+
+    def do_compress(self, filename, compressed_filename):
+        with open(filename, 'rb') as f_in, open(compressed_filename, 'wb') as f_out:
+            f_out.write(brotli.compress(f_in.read(), mode=brotli.MODE_TEXT))
+
+
+def get_compressor(settings):
+    name = settings.get('method', '')
+    if name == 'gzip':
+        return GZipCompressor(settings)
+    elif name == 'zopfli':
+        try:
+            global zopfli
```

The use of a global is to make it available in the `do_compress` method ? That's a bit weird, you can simply import it again in `do_compress`.

---

Yes, I have tried different things before doing that, but I'm not even convinced myself.
The goal is to fail fast, and to import once. I don't really like calling import on every file compression, even if it's fast.
Maybe the __init__ of each class do this job?